File: ios/ForkCallIssueState.m
// FORK — KAR-857
// Owns: Per-call connection-issue state and connected/issue sound transitions.
// Hooks into: TwilioVoiceReactNative+CallKit.m and ForkCallSounds.
// Re-check on SDK bump: Call quality warning names and reconnect callback ordering.

@import TwilioVoice;

#import "ForkCallIssueState.h"
#import "ForkCallSounds.h"

typedef NS_ENUM(NSInteger, ForkIssueState) {
    ForkIssueStateNoIssues,
    ForkIssueStateHasIssues,
};

@interface ForkCallIssueRecord : NSObject

@property (nonatomic, strong) NSUUID *uuid;
@property (nonatomic) BOOL reconnecting;
@property (nonatomic) BOOL poorNetworkQuality;
@property (nonatomic) NSTimeInterval nextSoundAllowedAt;
@property (nonatomic) ForkIssueState issueState;

@end

@implementation ForkCallIssueRecord
@end

static const NSTimeInterval ForkSoundThrottleSeconds = 2.0;
static const NSTimeInterval ForkEndedCallMarkerLifetimeSeconds = 30.0;
static NSMutableDictionary<NSString *, ForkCallIssueRecord *> *sStates;
static NSMutableSet<NSString *> *sRecentlyEndedCallUUIDs;

@implementation ForkCallIssueState

+ (void)fork_connectedCall:(NSUUID *)uuid {
    [self fork_onMainQueue:^{
        if ([self fork_hasRecentlyEnded:uuid]) return;

        ForkCallIssueRecord *record = [ForkCallIssueRecord new];
        record.uuid = uuid;
        record.issueState = ForkIssueStateNoIssues;
        [self fork_states][uuid.UUIDString] = record;
        [self fork_playIfAllowed:record sound:^{
            [ForkCallSounds fork_playConnectedSoundForCall:uuid];
        }];
    }];
}

+ (void)fork_reconnectingCall:(NSUUID *)uuid {
    [self fork_onMainQueue:^{
        ForkCallIssueRecord *record = [self fork_states][uuid.UUIDString];
        if (!record) return;

        record.reconnecting = YES;
        [self fork_updateRecord:record];
    }];
}

+ (void)fork_reconnectedCall:(NSUUID *)uuid {
    [self fork_onMainQueue:^{
        ForkCallIssueRecord *record = [self fork_states][uuid.UUIDString];
        if (!record) return;

        record.reconnecting = NO;
        [self fork_updateRecord:record];
    }];
}

+ (void)fork_call:(NSUUID *)uuid qualityWarningsChanged:(NSSet<NSNumber *> *)currentWarnings {
    BOOL poorNetworkQuality = [self fork_hasNetworkQualityWarning:currentWarnings];
    [self fork_onMainQueue:^{
        ForkCallIssueRecord *record = [self fork_states][uuid.UUIDString];
        if (!record) return;

        record.poorNetworkQuality = poorNetworkQuality;
        [self fork_updateRecord:record];
    }];
}

+ (void)fork_endedCall:(NSUUID *)uuid {
    [self fork_markRecentlyEnded:uuid];
    [self fork_onMainQueue:^{
        [[self fork_states] removeObjectForKey:uuid.UUIDString];
        [ForkCallSounds fork_stopConnectionStatusSoundForCall:uuid];
        [self fork_scheduleEndedMarkerRemoval:uuid];
    }];
}

+ (void)fork_onMainQueue:(dispatch_block_t)block {
    if ([NSThread isMainThread]) {
        block();
    } else {
        dispatch_async(dispatch_get_main_queue(), block);
    }
}

+ (NSMutableDictionary<NSString *, ForkCallIssueRecord *> *)fork_states {
    if (!sStates) sStates = [NSMutableDictionary dictionary];
    return sStates;
}

+ (BOOL)fork_hasNetworkQualityWarning:(NSSet<NSNumber *> *)warnings {
    return [warnings containsObject:@(TVOCallQualityWarningHighRtt)]
        || [warnings containsObject:@(TVOCallQualityWarningHighJitter)]
        || [warnings containsObject:@(TVOCallQualityWarningHighPacketsLostFraction)]
        || [warnings containsObject:@(TVOCallQualityWarningLowMos)];
}

+ (void)fork_updateRecord:(ForkCallIssueRecord *)record {
    ForkIssueState nextIssueState = record.reconnecting || record.poorNetworkQuality
        ? ForkIssueStateHasIssues
        : ForkIssueStateNoIssues;
    if (record.issueState == nextIssueState) return;

    record.issueState = nextIssueState;
    if (nextIssueState == ForkIssueStateHasIssues) {
        [self fork_playIfAllowed:record sound:^{
            [ForkCallSounds fork_playHasIssuesSoundForCall:record.uuid];
        }];
    } else {
        [self fork_playIfAllowed:record sound:^{
            [ForkCallSounds fork_playConnectedSoundForCall:record.uuid];
        }];
    }
}

+ (void)fork_playIfAllowed:(ForkCallIssueRecord *)record sound:(dispatch_block_t)playSound {
    NSTimeInterval now = [NSProcessInfo processInfo].systemUptime;
    if (now < record.nextSoundAllowedAt) return;

    record.nextSoundAllowedAt = now + ForkSoundThrottleSeconds;
    playSound();
}

+ (void)fork_markRecentlyEnded:(NSUUID *)uuid {
    @synchronized(self) {
        if (!sRecentlyEndedCallUUIDs) sRecentlyEndedCallUUIDs = [NSMutableSet set];
        [sRecentlyEndedCallUUIDs addObject:uuid.UUIDString];
    }
}

+ (BOOL)fork_hasRecentlyEnded:(NSUUID *)uuid {
    @synchronized(self) {
        return [sRecentlyEndedCallUUIDs containsObject:uuid.UUIDString];
    }
}

+ (void)fork_scheduleEndedMarkerRemoval:(NSUUID *)uuid {
    dispatch_after(
        dispatch_time(DISPATCH_TIME_NOW, (int64_t)(ForkEndedCallMarkerLifetimeSeconds * NSEC_PER_SEC)),
        dispatch_get_main_queue(),
        ^{
            @synchronized(self) {
                [sRecentlyEndedCallUUIDs removeObject:uuid.UUIDString];
            }
        }
    );
}

@end

File: android/src/main/java/com/twiliovoicereactnative/ForkCallIssueState.java
// FORK — KAR-857
// Owns: Per-call connection-issue state and connected/issue sound transitions.
// Hooks into: CallListenerProxy and ForkCallSounds.
// Re-check on SDK bump: Call quality warning names and reconnect callback ordering.
package com.twiliovoicereactnative;

import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;

import com.twilio.voice.Call;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

final class ForkCallIssueState {
  private static final long SOUND_THROTTLE_MILLIS = 2_000;
  private static final long ENDED_CALL_MARKER_LIFETIME_MILLIS = 30_000;
  private static final Handler mainHandler = new Handler(Looper.getMainLooper());
  private static final Map<UUID, State> states = new HashMap<>();
  private static final Set<UUID> recentlyEndedCalls = ConcurrentHashMap.newKeySet();

  private enum IssueState {
    NO_ISSUES,
    HAS_ISSUES
  }

  private static final class State {
    private boolean reconnecting;
    private boolean poorNetworkQuality;
    private long nextSoundAllowedAtMillis;
    private IssueState issueState = IssueState.NO_ISSUES;
  }

  private ForkCallIssueState() {}

  static void connected(UUID uuid) {
    runOnMainThread(() -> {
      if (recentlyEndedCalls.contains(uuid)) return;

      State state = new State();
      states.put(uuid, state);
      playIfAllowed(state, ForkCallSounds::playConnectedSound);
    });
  }

  static void reconnecting(UUID uuid) {
    runOnMainThread(() -> {
      State state = states.get(uuid);
      if (state == null) return;

      state.reconnecting = true;
      update(state);
    });
  }

  static void reconnected(UUID uuid) {
    runOnMainThread(() -> {
      State state = states.get(uuid);
      if (state == null) return;

      state.reconnecting = false;
      update(state);
    });
  }

  static void qualityWarningsChanged(
    UUID uuid,
    Set<Call.CallQualityWarning> currentWarnings
  ) {
    boolean poorNetworkQuality = hasNetworkQualityWarning(currentWarnings);
    runOnMainThread(() -> {
      State state = states.get(uuid);
      if (state == null) return;

      state.poorNetworkQuality = poorNetworkQuality;
      update(state);
    });
  }

  static void ended(UUID uuid) {
    recentlyEndedCalls.add(uuid);
    runOnMainThread(() -> {
      states.remove(uuid);
      ForkCallSounds.stopConnectionStatusSound();
      mainHandler.postDelayed(
        () -> recentlyEndedCalls.remove(uuid),
        ENDED_CALL_MARKER_LIFETIME_MILLIS
      );
    });
  }

  private static void runOnMainThread(Runnable operation) {
    if (Looper.myLooper() == Looper.getMainLooper()) {
      operation.run();
    } else {
      mainHandler.post(operation);
    }
  }

  private static boolean hasNetworkQualityWarning(
    Set<Call.CallQualityWarning> warnings
  ) {
    return warnings.contains(Call.CallQualityWarning.WARN_HIGH_RTT)
      || warnings.contains(Call.CallQualityWarning.WARN_HIGH_JITTER)
      || warnings.contains(Call.CallQualityWarning.WARN_HIGH_PACKET_LOSS)
      || warnings.contains(Call.CallQualityWarning.WARN_LOW_MOS);
  }

  private static void update(State state) {
    IssueState nextIssueState = state.reconnecting || state.poorNetworkQuality
      ? IssueState.HAS_ISSUES
      : IssueState.NO_ISSUES;
    if (state.issueState == nextIssueState) return;

    state.issueState = nextIssueState;
    if (nextIssueState == IssueState.HAS_ISSUES) {
      playIfAllowed(state, ForkCallSounds::playHasIssuesSound);
    } else {
      playIfAllowed(state, ForkCallSounds::playConnectedSound);
    }
  }

  private static void playIfAllowed(State state, Runnable playSound) {
    long now = SystemClock.elapsedRealtime();
    if (now < state.nextSoundAllowedAtMillis) return;

    state.nextSoundAllowedAtMillis = now + SOUND_THROTTLE_MILLIS;
    playSound.run();
  }
}
